### PR TITLE
Update Artifact Actions to use > v4.0.0

### PR DESCRIPTION
**Pull request by ashconnor**
### Justification

[update actions/download-artifact used by hashicorp/actions-docker-build](https://github.com/hashicorp/releng-support/issues/298)

### Summary

- The current major version v3 of the artifact upload/download actions uses a deprecated version of NodeJS (v16). This commit updates the artifact actions to use v4 which includes a supported version of NodeJS (v20).

### Quality

All changes to behavior should be accompanied by relevant tests which both document and protect it.

This PR includes:

  - [X] New or updated tests which validate the new behavior.  _(Thank you!)_
  - [ ] New or updated behavior which has been manually tested. _(Please provide a link to relevant logs if this is the case.)_
  - [ ] New or updated behavior which is not testable in a reasonable time frame. _(Please ask for help if this is the case, more things are testable than we sometimes think!)_
  - [ ] No new or changed behavior.  _(Just documentation, configuration, or pure refactoring.)_

### CODE REVIEW

**Comment by alvin-huang** [comment]
We will need to update https://github.com/hashicorp/actions-docker-build/blob/main/.github/workflows/test.yml#L255 for the tests too :)

**Comment by ashconnor** [comment]
Parallel `overwrite: true` is causing issues as described in [this issue](https://github.com/actions/upload-artifact/issues/506)

tl;dr: `overwrite` deletes the artifact and reuploads it. When done in parallel this can result in a 409 conflict.

**Comment by glennsarti** [comment]
Is it possible to get this PR merged and released soon? We don't have long until node16 support is dropped (may 13)

File: action.yml [hunk 1]
```diff
@@ -207,38 +207,46 @@ runs:
       env:
         IMAGE_NAME: ${{env.AUTO_TAG}}
 
+    - name: Generate Random String to use for tag list json
```
Comment (alvin-huang) [inline action.yml:210]: Is it easier to use the `GITHUB_RUN_ID` instead? https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables

Comment (alvin-huang) [reply action.yml:210]: Chatted about this in slack, that ID is the same across multiple matrix jobs so it's not unique

File: action.yml [hunk 2]
```diff
@@ -207,38 +207,46 @@ runs:
...
-      uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
+      uses: actions/upload-artifact@5d5d22a31266ced268874388b861e4b58bb5c2f3 # v4.3.1
       env:
         # Add _redhat if this is a redhat call.
         REDHAT_SUFFIX: ${{ inputs.redhat_tag && '_redhat' || '' }}
+        RANDOM: "${{ env.RANDOM_STRING }}"
       with:
-        name: docker_tag_list_${{env.TARGET}}${{env.REDHAT_SUFFIX}}.json
-        path: docker_tag_list_${{env.TARGET}}${{env.REDHAT_SUFFIX}}.json
+        name: docker_tag_list_${{env.TARGET}}${{env.REDHAT_SUFFIX}}.${{ inputs.arch }}.${{ env.RANDOM_STRING }}.json
```
Comment (alvin-huang) [inline action.yml:228]: Should this be `env.RANDOM`?

Comment (sarahethompson) [reply action.yml:228]: it should be` env.RANDOM_STRING`

File: action.yml [hunk 3]
```diff
@@ -207,38 +207,46 @@ runs:
...
+        RANDOM_STRING: "${{ env.RANDOM_STRING }}"
       run: $GITHUB_ACTION_PATH/scripts/create_metadata
 
     - name: Upload Docker Image metadata
-      uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
+      uses: actions/upload-artifact@5d5d22a31266ced268874388b861e4b58bb5c2f3 # v4.3.1
       env:
         # Add _redhat if this is a redhat call.
         REDHAT_SUFFIX: ${{ inputs.redhat_tag && '_redhat' || '' }}
+        RANDOM: "${{ env.RANDOM_STRING }}"
```
Comment (alvin-huang) [inline action.yml:226]: Do we need this if we don't use it?

File: .github/workflows/example.yml [hunk 4]
```diff
@@ -41,11 +41,12 @@ jobs:
           go build -o "$product_name" .
           zip "${{ env.product_name }}_${{ env.product_version }}_${{ matrix.goos }}_${{ matrix.goarch }}.zip" "$product_name"
       - name: Upload product artifact.
-        uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
+        uses: actions/upload-artifact@5d5d22a31266ced268874388b861e4b58bb5c2f3 # v4.3.1
         with:
           path: example/${{ env.product_name }}_${{ env.product_version }}_${{ matrix.goos }}_${{ matrix.goarch }}.zip
           name: ${{ env.product_name }}_${{ env.product_version }}_${{ matrix.goos }}_${{ matrix.goarch }}.zip
           if-no-files-found: error
+          overwrite: true
```
Comment (alvin-huang) [inline .github/workflows/example.yml:49]: Do we still need the overwrite here?

**Review by alvin-huang — APPROVED** [review]

**Comment by glennsarti** [comment]
Wohoo! Thankyou
